sklearn/linear_model/cd_fast.pyx

```diff
@@ -302,6 +303,13 @@ def enet_coordinate_descent(np.ndarray[floating, ndim=1] w,
                 if gap < tol:
                     # return if we reached desired tolerance
                     break
+                else:
```

Comment:
Shouldn't this be an else clause of the for loop, not of the if statement?

Please add a test that no warning is raised if the optimisation reaches convergence

Reply:
Since the function returns gap, can't we do this outside of the function??

Reply:
Issue #10813 suggested raising the warning when the max number of iterations is reached and the desired tolerance has yet to be achieved. It's not obvious to me how to test for that outside of the function because it doesn't return max_iter.

Should this be changed to look for instances where gap and tolerance are both equal to zero - indicating possible numerical issues?

Reply:
max_iter is passed in, so surely it is available to the caller test